cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -397,6 +397,130 @@ struct PowerChecked {
   }
 };
 
+// Bitwise operations
+
+struct BitWiseNot {
+  template <typename T, typename Arg>
+  static T Call(KernelContext*, Arg arg, Status*) {
+    return ~arg;
+  }
+};
+
+struct BitWiseAnd {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    return lhs & rhs;
+  }
+};
+
+struct BitWiseOr {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    return lhs | rhs;
+  }
+};
+
+struct BitWiseXor {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    return lhs ^ rhs;
+  }
+};
+
+struct ShiftLeftLogical {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    return lhs << rhs;
+  }
+};
+
+// See SEI CERT C Coding Standard rule INT34-C
+struct ShiftLeftLogicalChecked {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_unsigned_integer<T> Call(KernelContext*, Arg0 lhs, Arg1 rhs,
+                                            Status* st) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE(rhs < 0)) {
+      *st = Status::Invalid("Both operands must be non-negative");
+      return lhs;
+    }
+    if (ARROW_PREDICT_FALSE(rhs >= std::numeric_limits<Arg0>::digits)) {
+      *st = Status::Invalid("overflow");
+      return lhs;
+    }
+    return lhs << rhs;
+  }
+
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_signed_integer<T> Call(KernelContext*, Arg0 lhs, Arg1 rhs,
+                                          Status* st) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE(lhs < 0 || rhs < 0)) {
+      *st = Status::Invalid("Both operands must be non-negative");
```

Comment:
I don't understand why `lhs` cannot be negative? Left-shifting a negative number works fine (unless the value overflows, of course).

Reply:
A left shift is undefined on a negative number.

> if E1 has a signed type and non-negative value, and E1×2^E2 is representable in the corresponding unsigned type of the result type, then that value, converted to the result type, is the resulting value; otherwise, the behavior is undefined

[C++11 standard 5.8.2 (page 118)](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf) or see the [MSVC page](https://docs.microsoft.com/en-us/cpp/cpp/left-shift-and-right-shift-operators-input-and-output?view=msvc-160#left-shifts)

Reply:
Yawn. The C++ spec is completely stupid on this (just like the C spec).

Left-shifting of negative numbers is well-defined in Python:
```python
>>> -5 << 1
-10
>>> -5 << 3
-40
```
Also in Java, where the definition is based on two's complement representation:
https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.19

The Rust reference doesn't say anything about left-shifting of negative numbers (either [here](https://doc.rust-lang.org/reference/expressions/operator-expr.html#arithmetic-and-logical-binary-operators) or [there](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)), but a quick test seems to work.

Reply:
We could cast to and from unsigned to do the shift; that would sidestep the undefinedness and give the equivalent result (assuming a two's complement machine which I think we are ok assuming).

Reply:
In any case, I think we should support a well-defined, obvious, definition of left-shifting negative numbers just like most modern programming languages. The implementation should just be a matter of using the corresponding unsigned operation, AFAICT.

Reply:
Should we also explicitly mask the shift amount (as Java specifies, and as x86 does - ARM does not)? Then the only check would be effectively that the shift amount is not negative.

Reply:
There seems be no consensus among SQL databases. Let's err on the side of caution and return an error if the shift value is out bounds (including negative).